File: LeetCode/Hard/1900-The_Earliest_And_Latest_Rounds_Where_Players_Compete/1900-TheEarliestAndLatestRoundsWherePlayersCompete.cpp
#include <vector>
#include <climits>

using namespace std;

class Solution
{
private:
    bool memo[28][28][28];
    int min_round = INT_MAX;
    int max_round = INT_MIN;
    int n;

public:
    void solve(int alive, int p1, int p2, int l, int r, int round, int l_count, int m_count, int r_count)
    {
        if (l >= r)
            solve(alive, p1, p2, 0, n - 1, round + 1, l_count, m_count, r_count);
        else if ((alive & (1 << l)) == 0)
            solve(alive, p1, p2, l + 1, r, round, l_count, m_count, r_count);
        else if ((alive & (1 << r)) == 0)
            solve(alive, p1, p2, l, r - 1, round, l_count, m_count, r_count);
        else if (l == p1 && r == p2)
        {
            min_round = min(min_round, round);
            max_round = max(max_round, round);
        }
        else if (!memo[l_count][m_count][r_count])
        {
            memo[l_count][m_count][r_count] = true;

            if (l != p1 && l != p2)
            {
                solve(
                    alive ^ (1 << l),
                    p1,
                    p2,
                    l + 1,
                    r - 1,
                    round,
                    l_count - (l < p1),
                    m_count - (l > p1 && l < p2),
                    r_count - (l > p2));
            }

            if (r != p1 && r != p2)
            {
                solve(
                    alive ^ (1 << r),
                    p1,
                    p2,
                    l + 1,
                    r - 1,
                    round,
                    l_count - (r < p1),
                    m_count - (r > p1 && r < p2),
                    r_count - (r > p2));
            }
        }
    }
    vector<int> earliestAndLatest(int n, int firstPlayer, int secondPlayer)
    {
        firstPlayer -= 1;
        secondPlayer -= 1;
        this->n = n;
        int alive = (1 << n) - 1;
        memset(memo, false, sizeof(memo));

        solve(
            alive,
            firstPlayer,
            secondPlayer,
            0,
            n - 1,
            1,
            firstPlayer,
            secondPlayer - firstPlayer - 1,
            n - 1 - secondPlayer);

        return {min_round, max_round};
    }
};

File: LeetCode/Hard/1900-The_Earliest_And_Latest_Rounds_Where_Players_Compete/1900-TheEarliestAndLatestRoundsWherePlayersCompete.py
class Solution:
    def earliestAndLatest(
        self, n: int, firstPlayer: int, secondPlayer: int
    ) -> list[int]:
        firstPlayer -= 1
        secondPlayer -= 1
        min_round = float("inf")
        max_round = float("-inf")
        alive_mask = (1 << n) - 1
        memo = [[[False] * 28 for _ in range(28)] for _ in range(28)]

        def solve(alive, l, r, round, left_count, mid_count, right_count):
            if l >= r:
                # Next rount
                solve(alive, 0, n - 1, round + 1, left_count, mid_count, right_count)
            elif alive & (1 << l) == 0:
                # If `l` is not alive
                solve(alive, l + 1, r, round, left_count, mid_count, right_count)
            elif alive & (1 << r) == 0:
                # If `r` is not alive
                solve(alive, l, r - 1, round, left_count, mid_count, right_count)
            elif l == firstPlayer and r == secondPlayer:
                # If we found the match P1 vs P2
                # Update the max and min
                nonlocal min_round, max_round
                min_round = min(min_round, round)
                max_round = max(max_round, round)
            elif not memo[left_count][mid_count][right_count]:
                # If not calculated yet

                # Mark as calculated
                memo[left_count][mid_count][right_count] = True

                # Let r wins
                if l != firstPlayer and l != secondPlayer:
                    solve(
                        alive ^ (1 << l),
                        l + 1,
                        r - 1,
                        round,
                        left_count - (l < firstPlayer),
                        mid_count - (l > firstPlayer and l < secondPlayer),
                        right_count - (l > secondPlayer),
                    )

                # Let l wins
                if r != firstPlayer and r != secondPlayer:
                    solve(
                        alive ^ (1 << r),
                        l + 1,
                        r - 1,
                        round,
                        left_count - (r < firstPlayer),
                        mid_count - (r > firstPlayer and r < secondPlayer),
                        right_count - (r > secondPlayer),
                    )

        solve(
            alive_mask,
            0,
            0,
            0,
            firstPlayer,
            n - 1 - secondPlayer,
            secondPlayer - firstPlayer - 1,
        )

        return [min_round, max_round]

    def earliestAndLatest1(
        self, n: int, firstPlayer: int, secondPlayer: int
    ) -> list[int]:
        firstPlayer -= 1
        secondPlayer -= 1
        min_round = float("inf")
        max_round = float("-inf")
        alive = [True] * n
        memo = [[[False] * 28 for _ in range(28)] for _ in range(28)]

        def solve(alive, l, r, round, left_count, mid_count, right_count):
            if l >= r:
                # Next rount
                solve(alive, 0, n - 1, round + 1, left_count, mid_count, right_count)
            elif not alive[l]:
                # If `l` is not alive
                solve(alive, l + 1, r, round, left_count, mid_count, right_count)
            elif not alive[r]:
                # If `r` is not alive
                solve(alive, l, r - 1, round, left_count, mid_count, right_count)
            elif l == firstPlayer and r == secondPlayer:
                # If we found the match P1 vs P2
                # Update the max and min
                nonlocal min_round, max_round
                min_round = min(min_round, round)
                max_round = max(max_round, round)
            elif not memo[left_count][mid_count][right_count]:
                # If not calculated yet

                # Mark as calculated
                memo[left_count][mid_count][right_count] = True

                copy = alive.copy()
                # Let r wins
                if l != firstPlayer and l != secondPlayer:
                    copy[l] = False
                    solve(
                        copy,
                        l + 1,
                        r - 1,
                        round,
                        left_count - (l < firstPlayer),
                        mid_count - (l > firstPlayer and l < secondPlayer),
                        right_count - (l > secondPlayer),
                    )
                    copy[l] = True

                # Let l wins
                if r != firstPlayer and r != secondPlayer:
                    copy[r] = False
                    solve(
                        copy,
                        l + 1,
                        r - 1,
                        round,
                        left_count - (r < firstPlayer),
                        mid_count - (r > firstPlayer and r < secondPlayer),
                        right_count - (r > secondPlayer),
                    )

        solve(
            alive,
            0,
            0,
            0,
            firstPlayer,
            n - 1 - secondPlayer,
            secondPlayer - firstPlayer - 1,
        )

        return [min_round, max_round]


a = Solution()
print(a.earliestAndLatest(n=11, firstPlayer=2, secondPlayer=4))
print(a.earliestAndLatest(n=5, firstPlayer=1, secondPlayer=5))
